_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,375 @@
+
+In case there is no return value present, the object's equality cannot be asserted. Here Mockito's Spy comes into play.
+The test double can be queried, if a method was called with the expected parameter. This is what `Mockito.verify()`
+does. Those two test cases are widely used and give you a huge possibility to create various simple test cases. Also,
+the shown examples can be used for test driven development and regression tests. Mockito fits both needs!
+
+## How to Create Mocks with Mockito
+
+Until now, we have seen how to create quickly fast and simple test cases. Now let's face the possibilities to create the
+mocks for our needs.
+
+### Plain Mockito
+
+In case you don't want to use any framework nor annotations, you can create the mocks you need as follows:
+
+```java
+ClassToMock mock = Mockito.mock(ClassToMock.class);
+```
+
+That's all you need to create a mock with Mockito!
+
+### Mockito Annotation Style
+
+```java
+class CityServiceImplTestMockitoAnnotationStyle {
+
+    // System under Test (SuT)
+    private CityService cityService;
+
+    // Mocks
+    @Mock
+    private CityRepository cityRepository;
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
+        cityService = new CityServiceImpl(cityRepository);
+    }
+
+}
+```
+
+If the system under test has several dependencies that must be mocked, it gets cumbersome to create all these mocks with
+the variant shown above. You can annotate them with `@Mock` and initialize them all at once. Applying this variant, you
+don't have to deal with boilerplate code and are able to keep your unit test neat and concise.
+
+### Mockito JUnit Extension Style
+
+```java
+
+@ExtendWith(MockitoExtension.class)
+class CityServiceImplTestMockitoJUnitExtensionStyle {
+
+    // System under Test (SuT)
+    private CityService cityService;
+
+    // Mocks
+    @Mock
+    private CityRepository cityRepository;
+
+    @BeforeEach
+    void setUp() {
+        cityService = new CityServiceImpl(cityRepository);
+    }
+
+}
+```
+
+An alternative to the Mockito Annotation Style you can make use of JUnit Jupiter's `@ExtendWith` and extend the
+Jupiter's context with `MockitoExtension.class`.
+
+This makes your setup even neater and conciser!
+
+### Mockito Spring Style
+
+```java
+
+@ExtendWith(SpringExtension.class)
+class CityServiceImplTestMockitoSpringStyle {
+
+    // System under Test (SuT)
+    private CityService cityService;
+
+    // Mocks
+    @MockBean
+    private CityRepository cityRepository;
+
+    @BeforeEach
+    void setUp() {
+        cityService = new CityServiceImpl(cityRepository);
+    }
+
+}
+```
+
+If you have a more complex test fixture, and you want to inject the mock into Spring's `ApplicationContext` you can make
+use of `@MockBean`. But caution: `@MockBean` is not an annotation from Mockito but from Spring!
+
+## Mockito in Wilderness
```

Comment:
I would call this section "Mockito Best Practices" (against my advice for "active" section titles above). That will draw people's attentions to the section.